File: algorithms/svd_euclidean.py
from __future__ import print_function
import numpy as np
import sys

from algorithms.svd import SVD
from utils.data_io import get_user_movie_time_rating
import utils.c_interface


class SVDEuclidean(SVD):
    def initialize_users_and_movies(self):
        self.max_user = self.calculate_max_user()
        self.max_movie = self.calculate_max_movie()
        np.random.seed()
        self.users = np.array(
            np.random.normal(loc=0.0, scale=self.feature_initial,
                             size=(self.max_user, self.num_features)),
            dtype=np.float32)
        self.movies = np.array(
            np.random.normal(loc=0.0, scale=self.feature_initial,
                             size=(self.max_movie, self.num_features)),
            dtype=np.float32)

    def train(self, train_points, stats, epochs=1):
        self.set_train_points(train_points=train_points)
        self.set_stats(stats=stats)
        self.initialize_users_and_movies()
        for epoch in range(epochs):
            if self.debug:
                print('Epoch {}'.format(epoch+1))
                print('movies: {}'.format(self.movies))
                print('users: {}'.format(self.users))
            if np.isnan(np.sum(self.movies)) or np.isnan(np.sum(self.users)):
                print("So, I found a NaN..")
                import pdb
                pdb.set_trace()
            if self.run_c:
                self.train_epoch_in_c()
            else:
                self.train_epoch()

    def train_more(self, train_points=None, epochs=1):
        if train_points is not None:
            self.set_train_points(train_points)
        for epoch in range(epochs):
            if self.run_c:
                self.train_epoch_in_c()
            else:
                self.train_epoch()

    def train_epoch(self):
        count = 0
        for train_point in self.train_points:
            count += 1
            if count % 100000 == 0:
                sys.stdout.write('.')
                sys.stdout.flush()
            user, movie, _, rating = get_user_movie_time_rating(train_point)
            self.update_euclidean_all_features(user=user, movie=movie,
                                               rating=rating)

    def train_epoch_in_c(self):
        utils.c_interface.c_svd_euclidean_train_epoch(
            train_points=self.train_points,
            users=self.users,
            user_offsets=self.stats.user_offsets,
            movies=self.movies,
            movie_averages=self.stats.movie_averages,
            num_features=self.num_features,
            learn_rate=self.learn_rate,
            k_factor=self.k_factor
        )

    def update_euclidean_all_features(self, user, movie, rating):
        prediction_error = self.calculate_prediction_error(user=user,
                                                           movie=movie,
                                                           rating=rating)
        for feature in range(self.num_features):
            self.update_user_and_movie(user=user, movie=movie, feature=feature,
                                       error=prediction_error)

File: algorithms/svd_euclidean.c
#include <stdio.h>

int c_train_epoch(int *train_points, int num_points, float *users, float *user_offsets,
        int num_users, float *movies, float* movie_averages, int num_movies,
        float learn_rate, int num_features, float k_factor)
{

	int p, f;
	float prediction;
	float *user_features_cursor, *movie_features_cursor;
	float error, user_change, movie_change;
	int *train_cursor = train_points;
	int user_id, movie_id, time, rating;

	for(p = 0; p < num_points; p++) {
	    // get user id's locally
        //user_id = train_points[p*4];
        //movie_id = train_points[p*4+1];
        //rating = train_points[p*4+3];
		user_id    = *(train_cursor++);
		movie_id   = *(train_cursor++);
		time       = *(train_cursor++);
		rating     = *(train_cursor++);
        // Calculate the prediction error:
        // start prediction at baseline:
        prediction = movie_averages[movie_id] + user_offsets[user_id];
        // then: add features dot product to prediction
        user_features_cursor = users + user_id * num_features;
        movie_features_cursor = movies + movie_id * num_features;
        for (f = 0; f < num_features; f++) {
            prediction += user_features_cursor[f] * movie_features_cursor[f];
            if (prediction > 5) {
                prediction = 5.0;
            } else if (prediction < 1) {
                prediction = 1.0;
            }
        }

        // Calculate error:
        error = ((float) rating) - prediction;

        // Update the features
        // reset feature cursors
        for (f = 0; f < num_features; f++) {
            user_change = learn_rate * (error * movie_features_cursor[f] 
                - k_factor * user_features_cursor[f]);
            movie_features_cursor[f] += learn_rate * (error * user_features_cursor[f] 
                - k_factor * movie_features_cursor[f]);
            user_features_cursor[f] += user_change;
        }

	}
    return 0;
}


